src/display/svg.js

```diff
@@ -982,12 +982,65 @@ SVGGraphics = class SVGGraphics {
    */
   _makeColorN_Pattern(args) {
     if (args[0] === 'TilingPattern') {
-      warn('Unimplemented pattern TilingPattern');
-      return null;
+      return this._makeTilingPattern(args);
     }
     return this._makeShadingPattern(args);
   }
 
+  /**
+   * @private
+   */
+  _makeTilingPattern(args) {
+    const color = args[1];
+    const operatorList = args[2];
+    const matrix = args[3] || IDENTITY_MATRIX;
+    const [x0, y0, x1, y1] = args[4];
+    const xstep = args[5];
+    const ystep = args[6];
+    const paintType = args[7];
+
+    const tilingId = `shading${shadingCount++}`;
+    const [tx0, ty0] = Util.applyTransform([x0, y0], matrix);
+    const [tx1, ty1] = Util.applyTransform([x1, y1], matrix);
+    const [xscale, yscale] = Util.singularValueDecompose2dScale(matrix);
+    const txstep = xstep * xscale;
+    const tystep = ystep * yscale;
+
+    const tiling = this.svgFactory.createElement('svg:pattern');
+    tiling.setAttributeNS(null, 'id', tilingId);
+    tiling.setAttributeNS(null, 'patternUnits', 'userSpaceOnUse');
+    tiling.setAttributeNS(null, 'width', txstep);
+    tiling.setAttributeNS(null, 'height', tystep);
+    tiling.setAttributeNS(null, 'x', `${tx0}`);
+    tiling.setAttributeNS(null, 'y', `${ty0}`);
+
+    // Save current state.
+    const svg = this.svg;
+    const transformMatrix = this.transformMatrix;
+    const fillColor = this.current.fillColor;
+    const strokeColor = this.current.strokeColor;
+
+    const bbox = this.svgFactory.create(tx1 - tx0, ty1 - ty0);
+    this.svg = bbox;
+    this.transformMatrix = matrix;
+    if (paintType === 2) {
+      const cssColor = Util.makeCssRgb(...color);
+      this.current.fillColor = cssColor;
+      this.current.strokeColor = cssColor;
+    }
+    this.executeOpTree(this.convertOpList(operatorList));
+
+    // Restore saved state.
+    this.svg = svg;
+    this.transformMatrix = transformMatrix;
+    this.fillColor = fillColor;
+    this.strokeColor = strokeColor;
+
+    tiling.appendChild(bbox.childNodes[0]);
```

Comment:
This line appears to break the PDF file from #10487, which also uses tiling patterns. Could you check why that happens?

Reply:
The #10487 PDF uses the TilingPattern on the first page only.
Before implementing TilingPattern, paths that should be drawn in red were black with `fill=""`.
Looking at the output of this PR implementation, the rendering of `<svg:pattern>` using `<svg:g clip-path="...">` is broken.
I can infer from this that there might be a mistake in the implementation associated with clip, but I didn't know exactly what was wrong.

Reply:
I have been thinking about this some more and I think it's best to leave fixing this for a follow-up. The PDF file already wasn't rendering correctly and there is an open issue for it. Given that this patch improves the other PDF files, I don't see this as a blocker for landing this.